### DISCUSSION TITLE
Modernize serverless-sharp

### DISCUSSION THREAD

**Discussion by bs-thomas** [venveo/serverless-sharp#122] [Q&A]
I have pulled your latest "modernize" branch, which you're trying to upgrade to the latest serverless-sharp, and a number of other tweaks.

I tried deploying, but stuck with some error messages in serverless.yml.
May I ask if this is still a WIP branch?

There are also a number of things I noticed:

1) serverless deprecates the usage of settings.yml, and promotes usage of env files.  Would you considering going for the latter option?  I can assist maybe if you're ok?

2) I see there is some Makefile and stuff.  But the sharp layer for Lambda doesn't really require any building I think.
All I did on my end is:
- Add a package.json in the directory "\lambdaLayers\SharpLayer\nodejs" with dependency "sharp": "^0.30.6"
- Go to the directory, and type in the command `npm install --platform=linux --arch=x64 sharp`
- Then, all the linux x64 binaries will be there, and you just deploy over and it works for me.

Or did I miss anything here?

3) Switched over to use TS.  Actually I have no experience on this part.  Is it really better?

4) TLSv1 <-- still uses this protocol.  AWS seems to already recommend 'TLSv1.2_2021'

=====

Btw, @Mosnar any communication tools you use?  I think we can exchange contacts, it can make our communication easier, and also to make a new overseas friend ;-)  If you're interested of course.

**Comment by Mosnar** [comment]
This branch is definitely still in development, sorry for the confusion. I do appreciate the feedback though! I've added notes with your feedback to the to-do list in the PR. Feel free to reach out on Discord - I'd love to collaborate. `Ransom#8483`

**Comment by Mosnar** [comment]
> serverless deprecates the usage of settings.yml, and promotes usage of env files. Would you considering going for the latter option? I can assist maybe if you're ok?

I couldn't find any documentation about this - are you able to share a link? I think given the complexity of the configuration, I'd almost rather it be a JSON file or something.

> I see there is some Makefile and stuff. But the sharp layer for Lambda doesn't really require any building I think.
> All I did on my end is:
> 
> Add a package.json in the directory "\lambdaLayers\SharpLayer\nodejs" with dependency "sharp": "^0.30.6"
> Go to the directory, and type in the command npm install --platform=linux --arch=x64 sharp
> Then, all the linux x64 binaries will be there, and you just deploy over and it works for me.
> Or did I miss anything here?

The goal here is to make the build process more repeatable - the Makefile runs the build in a Docker container configured in a way that we know will work, ensuring we have all of the correct dependencies, node versions, etc. Most people are probably running a newer version of Node on their machine for local tasks, so this prevents forcing people to install another Node and potentially messing up their other development environments just to update Sharp. The Makefile also gives us the opportunity to configure additional parameters for building Sharp from scratch down the road.

> Switched over to use TS. Actually I have no experience on this part. Is it really better?

It's a million times better - especially in a project like this where there's so much data flow. Trust me. I found 5 or 6 bugs in the old code just from the type-checking. 

> TLSv1 <-- still uses this protocol. AWS seems to already recommend 'TLSv1.2_2021'

Good catch - we should fix that.

**Reply by bs-thomas** [reply]
Let's 

> > serverless deprecates the usage of settings.yml, and promotes usage of env files. Would you considering going for the latter option? I can assist maybe if you're ok?
> 
> I couldn't find any documentation about this - are you able to share a link? I think given the complexity of the configuration, I'd almost rather it be a JSON file or something.

Ah sorry, my wording is a bit misleading.  Let's call it this way, there has never been any native way of dealing with these environments.  We have to define this settings file, and tell serverless.yml to load it and parse it.  And some users didn't like that, and introduced a plugin for ENV.  But serverless 3 has natively supported this now, so it's a native environment solution.

https://www.serverless.com/framework/docs/environment-variables

ENV, IMO is way nicer because environments are specified in different files, and you can compare them side by side very easily using a merge tool.  Specific environments (eg. dev, stage) can be passed onto another colleague easily without accidentally passing production credentials.

Let me know what you think.  I actually have something working on my end, and I personally think it's a neater solution.

> 
> > I see there is some Makefile and stuff. But the sharp layer for Lambda doesn't really require any building I think.
> > All I did on my end is:
> > Add a package.json in the directory "\lambdaLayers\SharpLayer\nodejs" with dependency "sharp": "^0.30.6"
> > Go to the directory, and type in the command npm install --platform=linux --arch=x64 sharp
> > Then, all the linux x64 binaries will be there, and you just deploy over and it works for me.
> > Or did I miss anything here?
> 
> The goal here is to make the build process more repeatable - the Makefile runs the build in a Docker container configured in a way that we know will work, ensuring we have all of the correct dependencies, node versions, etc. Most people are probably running a newer version of Node on their machine for local tasks, so this prevents forcing people to install another Node and potentially messing up their other development environments just to update Sharp. The Makefile also gives us the opportunity to configure additional parameters for building Sharp from scratch down the road.

Understood.  It makes a lot of sense.

> 
> > Switched over to use TS. Actually I have no experience on this part. Is it really better?
> 
> It's a million times better - especially in a project like this where there's so much data flow. Trust me. I found 5 or 6 bugs in the old code just from the type-checking.

Ah I see, maybe indeed I should look deeper into this.

> 
> > TLSv1 <-- still uses this protocol. AWS seems to already recommend 'TLSv1.2_2021'
> 
> Good catch - we should fix that.

;-)
